#10596 - Remove fmt fortran workarounds - Z & N (and S)

jmarrec opened the pull request:
Pull request overview
---------------------

- Fixes #10596

### Description of the purpose of this PR

The goal of this PR:

 * Remove fortran formats one after another
 * Analyze diffs at each commit, adjust to minimize diff, and explain the changes
 * Rince and repeat without losing sanity


### Pull Request Author

<!-- Add to this list or remove from it as applicable.  This is a simple templated set of guidelines. -->

 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer

<!-- This will not be exhaustively relevant to every PR. -->

 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

jmarrec committed:
Remove the Z format

jmarrec committed:
Replace the N format with the G format

jmarrec committed:
When removing Z format, need to adjust one unit test

jmarrec committed:
Adjust former N replacement: for U-value/R-value, we expect NOT needing an exponent. The '{:15.4G}' format will format '0.8' as '0.8' while {:15.4F} will format at 0.8000

```
#include <fmt/format.h>

int main() {
    fmt::print("{{:15.4G}}='{:15.4G}'\n", 0.8);
    fmt::print("{{:15.4F}}='{:15.4F}'\n", 0.8);
}
```

Output:
```
{:15.4G}='            0.8'
{:15.4F}='         0.8000'
```

jmarrec commented on src/EnergyPlus/Construction.cc:
```diff
@@ -1900,7 +1900,7 @@ void ConstructionProps::calculateFinalCoefficients()
 {
 
-    static constexpr std::string_view Format_700{" Construction CTF,{},{:4},{:4},{:4},{:8.3F},{:15.4G},{:8.3F},{:8.3F},{:8.3F},{:8.3F},{}\n"};
+    static constexpr std::string_view Format_700{" Construction CTF,{},{:4},{:4},{:4},{:8.3F},{:15.4F},{:8.3F},{:8.3F},{:8.3F},{:8.3F},{}\n"};
```
Note about 9af23f3793bf4d3ec95e0cd1721f3d26320673dd which I'm going to revert

I wrote in the commit message: 

Adjust former N replacement: for U-value/R-value, we expect NOT needing an exponent. The '{:15.4G}' format will format '0.8' as '0.8' while {:15.4F} will format at 0.8000

```
#include <fmt/format.h>

int main() {
    fmt::print("{{:15.4G}}='{:15.4G}'\n", 0.8);
    fmt::print("{{:15.4F}}='{:15.4F}'\n", 0.8);
}
```

Output:
```
{:15.4G}='            0.8'
{:15.4F}='         0.8000'
```

This creates more problems that it solves actually. We do have very low values

![image](https://github.com/user-attachments/assets/3b76718d-90ec-42fb-9c92-b052f6e6af0a)


|          double           |    {:15.4G}     |    {:15.4F}     |
| ------------------------- | --------------- | --------------- |
|          1.5461211121e-05 |       1.546E-05 |          0.0000 |
|     0.0008155511551215415 |       0.0008156 |          0.0008 |
|        0.8155511551215415 |          0.8156 |          0.8156 |
|                       0.8 |             0.8 |          0.8000 |
|                  1.011212 |           1.011 |          1.0112 |
|                  176.9231 |           176.9 |        176.9231 |
|              29980.012323 |       2.998E+04 |      29980.0123 |

See
https://compiler-explorer.com/z/azE67q8a5

jmarrec committed:
Revert "Adjust former N replacement: for U-value/R-value, we expect NOT needing an exponent. The '{:15.4G}' format will format '0.8' as '0.8' while {:15.4F} will format at 0.8000"

This reverts commit 9af23f3793bf4d3ec95e0cd1721f3d26320673dd.
cf https://github.com/NREL/EnergyPlus/pull/11109#discussion_r2168430394 for why this is worse

jmarrec commented:
Ok, so now comparing only the N-format changes, from 61631632c77f847fa3dfb6e5938ffb47cdc7b88d to either d74b29f37ab71467f8a006b3c482ffb196ae6815 or f7dd5e7ef529f68bf35b15c09784d1fc181816c8 which are completely equivalent

I wrote a complete EIO parser that will parse the tables, in two formats: text, and str. (I realized in the process how broken the EIO is, I had to do all kinds of workarounds...)

TL;DR I have only TWO actual diffs, which are minimal

![image](https://github.com/user-attachments/assets/7aa813f3-6f90-45a6-b88f-ffc45b10502a)

The vast majority of changes are exponents, eg `0.81E-5 => 8.1E-5`

A fair number due to the fact that the G format will not display trailing zeros: for a value of 0.8, '{:15.4G}' will format as '0.8', and {:15.4N} would format as '0.8000'.

Here is a filterable HTML page for all diffs found (I removed duplicated values):

[N_format_filter.html.txt](https://github.com/user-attachments/files/20923330/N_format_filter.html.txt)

And here is a CSV (careful to treat "base" and "mod" as TEXT when importing it) and an ODS for the same

[N_format.csv](https://github.com/user-attachments/files/20923346/N_format.csv)
[N_format.ods](https://github.com/user-attachments/files/20923347/N_format.ods)

jmarrec committed:
Actually remove Z and N (and S) from IOFiles

jmarrec committed:
Missed a couple of weird formats

jmarrec committed:
Adjust Unit test: Note that In PreDefTableEntry I could format the E version with sigDigitsCounts - 1, but chose not to

jmarrec commented on src/EnergyPlus/OutputReportPredefined.cc:
```diff
@@ -1646,7 +1646,7 @@ namespace OutputReportPredefined {
             }
             state.dataOutRptPredefined->tableEntry(state.dataOutRptPredefined->numTableEntry).charEntry =
-                format("{:12.{}Z}", tableEntryReal, sigDigitCount);
+                format("{:12.{}E}", tableEntryReal, sigDigitCount);
```
@Myoldmopar so, the difference with the Z and E format right now is basically this (see the OutputReportTabular.unit.cc changes for more info https://github.com/NREL/EnergyPlus/pull/11109/files#diff-d541504a0ee106c6fbf7df90a2b6c096cbaf55db3f0f6be70ccd6dc5664950bdR9892 )

```
print("{:.3Z}", 123456789.1) => 0.123E+09
print("{:.3E}", 123456789.1) => 1.235E+08
```

So Z would have a leading zero, and 3 numbers. The E format doesn't use a leading zero, and adds 3 decimals after the floating point separator (unlike the G format which would use 3 significant numbers, period)



So to limit the changes here, I could do `sigDigitCount - 1` in that format call... But I think I'd rather not, as this takes the same amount of space in the HTML table and provides more accuracy. **What do you think?**

```
print("{:.2E}", 123456789.1) => 1.23E+08
```

nrel-bot-2 commented:
@jmarrec @myoldmopar it has been 36 days since this pull request was last updated.

jmarrec committed:
Replace the S format with G

jmarrec commented:
@Myoldmopar I rebased onto develop. I updated my eio_differ.py Gist https://gist.github.com/jmarrec/583a4b2dc5a38ce106afe609590f816c to include the EIO fix PR I made already.

Same results.

python eio_differ.py --base-testfiles-dir testfiles-25.2.0-58430fc53b --mod-testfiles-dir testfiles-25.2.0-9516dbbe69 --output-html --out-dir eio_diffs_pr11109

| idf_file                                | table_name           | column                     | key           |    base |     mod |   base_value |   mod_value | equal   | exponential   |
|:----------------------------------------|:---------------------|:---------------------------|:--------------|--------:|--------:|-------------:|------------:|:--------|:--------------|
| 5ZoneVAV-ChilledWaterStorage-Stratified | Material CTF Summary | ThermalResistance {m2-K/w} | Roof Membrane | 0.05938 | 0.05937 |      0.05938 |     0.05937 | False   | True          |
| HospitalLowEnergy                       | Material CTF Summary | ThermalResistance {m2-K/w} | 1/2IN Gypsum  | 0.07937 | 0.07938 |      0.07937 |     0.07938 | False   | True          

[eio_diffs_pr11109.zip](https://github.com/user-attachments/files/22136762/eio_diffs_pr11109.zip)
